### PR TITLE
ADDensity can act on undisplaced meshes

### PR DESCRIPTION
**Pull request by recuero**
The purpose of this PR is to allow `ADDensity` to act on undisplaced meshes by making displacement variables optional. No other changes intended.

 Ref. #15311

### CODE REVIEW

**Comment by recuero** [comment]
@dschwen

**Review by dschwen — DISMISSED** [review]

**Comment by tophmatthews** [comment]
I'm not totally sure about this...if acting on an undisplaced mesh, it can just get a `GenericConstantMaterial`, and it's pretty tragic if you accidentally didn't link displacements in Density

**Comment by tophmatthews** [comment]
image

**Comment by dschwen** [comment]
lol, thanks for posting the git blame :-D

**Comment by dschwen** [comment]
Points to consider

* We encourage putting `displacements` in the `[GlobalParams]` block, so if that practice is followed the parameter would always be captured.
* I think it is a good input syntax design philosophy to reduce the interdepencence of syntax. I.e. if one change is made do our best to avoid inducing a chain reaction of dependent changes. In this case if the user wants to test a simplified model without displacements, do not force them to also change the (very expressive) Density object.

**Comment by dschwen** [comment]
I do agree that we should catch the case where a displaced mesh exists but no displacements were specified. That just goes against the design of the Density class.

**Comment by tophmatthews** [comment]
Along the lines of `use_displaced_mesh=false` default, does it matter in ADDensity? In other words, is _grad_disp automatically set to zero in use_displaced_mesh? What does use_displaced_mesh impact? points?

**Comment by dschwen** [comment]
uhm wait, we need to work on the undisplaced mesh! otherwise the gradients of the displacement field are wrong!

**Comment by tophmatthews** [comment]
yeah...I'm a bit concerned about this for my own stuff, not sure why I didn't catch that before.

File: modules/misc/src/materials/ADDensity.C [hunk 1]
```diff
@@ -26,10 +26,14 @@ ADDensity::validParams()
 ADDensity::ADDensity(const InputParameters & parameters)
   : ADMaterial(parameters),
     _coord_system(getBlockCoordSystem()),
-    _disp_r(adCoupledValue("displacements", 0)),
+    _disp_r(coupledComponents("displacements") ? adCoupledValue("displacements", 0) : _ad_zero),
     _initial_density(getParam<Real>("density")),
     _density(declareADProperty<Real>("density"))
 {
+  if (getParam<bool>("use_displaced_mesh"))
+    mooseError("ADDensity needs to act on an undisplaced mesh. Use of a displaced mesh leads to "
```
Comment (dschwen) [inline modules/misc/src/materials/ADDensity.C:34]: please make this 

```
   paramError("use_displaced_mesh", "ADDensity needs to act on an undisplaced mesh. Use of a displaced mesh leads to incorrect gradient values");
```

**Review by dschwen — CHANGES_REQUESTED** [review]

File: modules/misc/src/materials/ADDensity.C (hunk 1 above)
Comment (tophmatthews) [reply modules/misc/src/materials/ADDensity.C:34]: You're kind of still circling...if displacements aren't provided, and `use_displaced_mesh=true`, you'll still get an undisplaced behavior. It seems to me:
- `use_displaced_mesh` defaults true
- if `use_displaced_mesh=true`, require displacements
- if u`se_displaced_mesh=false`, require displacements not to be coupled

Comment (tophmatthews) [reply modules/misc/src/materials/ADDensity.C:34]: would explicitly capture the behaviors you want with very clear input.

Comment (dschwen) [reply modules/misc/src/materials/ADDensity.C:34]: No. `use_displaced_mesh` must **never** be true! The key is to look if a displaced mesh _exists_!

Comment (dschwen) [reply modules/misc/src/materials/ADDensity.C:34]:
```
if (_fe_problem.getDisplacedProblem() && coupledComponents("displacements") == 0)
  paramError("displacements", "Displacement variables need to be specified if a displaced mesh exists.);
```

File: modules/misc/src/materials/ADDensity.C [hunk 2]
```diff
@@ -26,12 +26,21 @@ ADDensity::validParams()
...
 {
+  if (getParam<bool>("use_displaced_mesh"))
+    paramError("ADDensity needs to act on an undisplaced mesh. Use of a displaced mesh leads to "
+               "incorrect gradient values");
+
   // get coupled gradients
   const unsigned int ndisp = coupledComponents("displacements");
+
+  if (ndisp == 0 && _fe_problem.getDisplacedProblem())
+    mooseError(
```
Comment (dschwen) [inline modules/misc/src/materials/ADDensity.C:41]: paramError

Comment (recuero) [reply modules/misc/src/materials/ADDensity.C]: Thought this one would not qualify as such. Addressed.

File: modules/misc/src/materials/ADDensity.C [hunk 3]
```diff
@@ -26,12 +26,21 @@ ADDensity::validParams()
...
 {
+  if (getParam<bool>("use_displaced_mesh"))
+    paramError("ADDensity needs to act on an undisplaced mesh. Use of a displaced mesh leads to "
+               "incorrect gradient values");
+
   // get coupled gradients
   const unsigned int ndisp = coupledComponents("displacements");
+
+  if (ndisp == 0 && _fe_problem.getDisplacedProblem())
+    paramError(
```
Comment (dschwen) [inline modules/misc/src/materials/ADDensity.C:41]: paramError("displacements", "...");

Comment (dschwen) [reply modules/misc/src/materials/ADDensity.C:41]: First argument is the parameter name. It's missing here.

**Comment by dschwen** [comment]
That marmot failure was already fixed! WTH?!

**Comment by dschwen** [comment]
Looks like devel is still at 5/27. Which means no libmesh conda package, which means no passing marmot.

**Review by dschwen — APPROVED** [review]

**Pull request merged**